pkg/store/telemetry.go

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package store
+
+import (
+	"strconv"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+)
+
+// seriesStatsAggregator aggregates results from fanned-out queries into a histogram given their
+// response's shape.
+type seriesStatsAggregator struct {
+	queryDuration *prometheus.HistogramVec
+
+	seriesLeBuckets  []int64
+	samplesLeBuckets []int64
+	seriesStats      storepb.SeriesStatsCounter
+}
+
+// NewSeriesStatsAggregator is a constructor for seriesStatsAggregator.
+func NewSeriesStatsAggregator(
+	reg prometheus.Registerer,
+	durationQuantiles []float64,
+	sampleQuantiles []int64,
+	seriesQuantiles []int64,
+) *seriesStatsAggregator {
+	return &seriesStatsAggregator{
+		queryDuration: promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
+			Name:    "thanos_store_api_query_duration_seconds",
+			Help:    "Duration of the Thanos Store API select phase for a query.",
+			Buckets: durationQuantiles,
+		}, []string{"series_le", "samples_le"}),
+		seriesLeBuckets:  seriesQuantiles,
+		samplesLeBuckets: sampleQuantiles,
+		seriesStats:      storepb.SeriesStatsCounter{},
+	}
+}
+
+// Aggregate is an aggregator for merging `storepb.SeriesStatsCounter` for each incoming fanned out query.
+func (s *seriesStatsAggregator) Aggregate(stats storepb.SeriesStatsCounter) {
+	s.seriesStats.Series += stats.Series
```

Comment:
doesn't number of series overlap too?

Reply:
It does, but we are (or were back then) not tracking the what individual series was touched. It would be quite expensive to do the tracking.